### PR TITLE
Fixes retrials without Thread.sleep when thread gets interrupted

### LINKED ISSUE #624: Feign Default Retryer retries without sleep when thread gets interrupted
Hi,
I found the problem in method continueOrPropagate in class Retryer.Default. 
When the thread in which Retryer is running gets interrupted, Retryer performs retrying without sleep.
Flow looks like that:
1. Thread gets interrupted
2. The next attempt is performed if maxAttepts is not yet met
3. Thread should sleep for a given interval - but it's already interrupted so the InterruptedException is thrown
4. Current thread gets interrupted and retrying is performed again.

The following procedure is performed until the higher level methods/thread groups handle interruption which can take some time and lead to fleed of retrials.

In my opinion, this situation should be handled (for example by propagation of RetryableException).

```
public void continueOrPropagate(RetryableException e) {
    if (attempt++ >= maxAttempts) {
        throw e;
    }

    long interval;
    if (e.retryAfter() != null) {
        interval = e.retryAfter().getTime() - currentTimeMillis();
        if (interval > maxPeriod) {
            interval = maxPeriod;
        }
        if (interval < 0) {
            return;
        }
    } else {
        interval = nextMaxInterval();
    }
    try {
        Thread.sleep(interval);
    } catch (InterruptedException ignored) {
        Thread.currentThread().interrupt();
    }
    sleptForMillis += interval;
}
```

### LINKED ISSUE #643: Do a release
I will help do a release by the end of the week. Ping here PRs you think should make the cut. Ping also here if you want to audit the release!

### PR DESCRIPTION
**Pull request by MichalSzewczyk**
This PR aims to avoid flood of retries when thread, in which retryer is running, gets interrupted. Default retryer will rethrow FeignException in case of interrupted state.
Fixes #624 
Fixes #643

### CODE REVIEW

**Comment by velo** [comment]
This makes lots of sense to me, I just wonder if we can have a test somehow.

File: core/src/main/java/feign/Retryer.java [hunk 1]
```diff
@@ -75,6 +75,7 @@ public void continueOrPropagate(RetryableException e) {
         Thread.sleep(interval);
       } catch (InterruptedException ignored) {
         Thread.currentThread().interrupt();
+        throw e;
```
[the remark this answers is not in the document]
Comment (rage-shadowman) [reply core/src/main/java/feign/Retryer.java]: Also, how does this not result in a compile error?

`InterruptedException` is checked, and that method doesn't declare any thrown checked exceptions?

[the remark this answers is not in the document]
Comment (MichalSzewczyk) [reply core/src/main/java/feign/Retryer.java]: I'm not going to rethrow ```InterruptedException``` but ```RetryableException```.

Comment (rage-shadowman) [reply core/src/main/java/feign/Retryer.java]: Oh gawd I'm an idiot... `e` vs `ignored`...

**Comment by rage-shadowman** [comment]
Might I suggest a test:

```java
  @Test
  public void failsOnInterruptedException() throws Exception {
    Default retryer = new Retryer.Default();

    Thread.currentThread().interrupt();
    RetryableException expected = new RetryableException(null, null, new Date(5000));
    try {
      retryer.continueOrPropagate(expected);
      Assert.fail("Retryer continued despite interruption");
    } catch (RetryableException e) {
      Assert.assertEquals("Unexpected exception found", expected, e);
    }
  }
```

**Comment by MichalSzewczyk** [comment]
`new Date(5000)` could be replaced with `null`. 
Btw is it safe to interrupt current thread? `Thread.currentThread().interrupt();` code will set flag back to interrupted and as a consequence all tests run after this one will have interrupted flag set to true.

**Comment by MichalSzewczyk** [comment]
What do you think about running retryer in other thread?
```
  @Test
  public void defaultRetryerShouldPropagateFeignExceptionWhenThreadGetsInterrupted() {
    Retryer retryer = new Retryer.Default();
    Thread thread = new Thread(() -> assertThatThrowsException(RetryableException.class, () -> retryer.continueOrPropagate(new RetryableException(null, null))));
    thread.start();
    thread.interrupt();
  }
```

**Comment by rage-shadowman** [comment]
You're right. Since the code resets the interrupted flag, you need to check it immediately afterward again. My code missed that part.

**Comment by rage-shadowman** [comment]
PS - I think that if you run it in another thread, you need to check that thread for assertion states. It's probably easier to just do it in the current thread and just `Assert.assertTrue(Thread.currentThread().isInterrupted())`

File: core/src/test/java/feign/RetryerTest.java [hunk 2]
```diff
@@ -74,4 +77,23 @@ protected long currentTimeMillis() {
...
   }
+
+  @Test
+  public void defaultRetryerShouldPropagateFeignExceptionWhenThreadGetsInterrupted() {
+    Retryer retryer = new Retryer.Default();
+    Thread thread = new Thread(() ->
+            assertThatThrowsExceptionAndThreadIsInterrupted(RetryableException.class,
+                    () -> retryer.continueOrPropagate(new RetryableException(null, null))));
+    thread.start();
+    thread.interrupt();
```
Comment (rage-shadowman) [inline core/src/test/java/feign/RetryerTest.java]: How can this test fail with no assertions while swallowing all exceptions in another thread?

**Comment by MichalSzewczyk** [comment]
Ok, I added test suggested by you.

**Pull request merged**